bigtable_automation/gcf/private.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package private runs a GCF function that triggers in 2 scenarios:
+//
+// 1) completion of prophet-flume job in borg. On triggering it sets up new
+//    cloud BT table, scales up BT cluster (if needed) and starts a dataflow job.
+// 2) completion of BT cache ingestion dataflow job. It scales BT cluster down
+//    (if needed).
+// The trigger function BTImportController requires a set of environment
+// variables to be set. They are stored in *.yaml files.
+
+package gcf
+
+import (
+	"context"
+	"log"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+func privateInternal(ctx context.Context, e GCSEvent) error {
+	projectID := os.Getenv("projectID")
+	bucket := os.Getenv("bucket")
+	instance := os.Getenv("instance")
+	cluster := os.Getenv("cluster")
+	nodesHigh := os.Getenv("nodesHigh")
+	nodesLow := os.Getenv("nodesLow")
+	dataflowTemplate := os.Getenv("dataflowTemplate")
+	if projectID == "" {
+		return errors.New("projectID is not set in environment")
+	}
+	if instance == "" {
+		return errors.New("instance is not set in environment")
+	}
+	if cluster == "" {
+		return errors.New("cluster is not set in environment")
+	}
+	if dataflowTemplate == "" {
+		return errors.New("dataflowTemplate is not set in environment")
+	}
+	// Get low and high nodes number
+	nodesH, err := strconv.Atoi(nodesHigh)
+	if err != nil {
+		return errors.Wrap(err, "Unable to parse 'nodesHigh' as an integer")
+	}
+	nodesL, err := strconv.Atoi(nodesLow)
+	if err != nil {
+		return errors.Wrap(err, "Unable to parse 'nodesLow' as an integer")
+	}
+	// Get table ID.
+	// e.Name should is like "**/<user>/<import>/control/<table_id>/launched.txt"
+	parts := strings.Split(e.Name, "/")
+	if parts[len(parts)-3] != "control" {
```

Comment:
might be more readable to use tmp vars?
idxControl := len(parts) - 3
idxTable := len(parts) -2